function-controller/pkg/controller/autoscaler/autoscaler.go

```diff
@@ -0,0 +1,325 @@
+	a.mutex.Lock()
+	defer a.mutex.Unlock()
+	funcTotals, ok := a.totals[topic]
+	if !ok {
+		funcTotals = make(map[FunctionId]*metricsTotals)
+		a.totals[topic] = funcTotals
+	}
+
+	_, ok = funcTotals[fn]
+	if ok {
+		return fmt.Errorf("Already monitoring topic %s and function %s", topic, fn)
+	}
+
+	funcTotals[fn] = &metricsTotals{}
+
+	a.scalers[fn] = decorate(a.metricsScaler(fn), a.limitScalingUp(fn), a.limitScalingDown(fn), a.delay(fn))
+
+	return nil
+}
+
+func (a *autoScaler) delay(fn FunctionId) adjuster {
+	p := NewDelayer(func() time.Duration {
+		return a.delayScaleDown(fn);
+	})
+
+	return func(proposal int) int {
+		return p.Delay(proposal).Get()
+	}
+}
+
+func (a *autoScaler) StopMonitoring(topic string, function FunctionId) error {
+	a.mutex.Lock()
+	defer a.mutex.Unlock()
+	funcTotals, ok := a.totals[topic]
+	if !ok {
+		return fmt.Errorf("Not monitoring topic %s and function %s", topic, function)
+	}
+
+	_, ok = funcTotals[function]
+	if !ok {
+		return fmt.Errorf("Not monitoring topic %s and function %s", topic, function)
+	}
+
+	delete(funcTotals, function)
+
+	// Avoid leaking memory.
+	if len(funcTotals) == 0 {
+		delete(a.totals, topic)
+	}
+	delete(a.scalers, function)
+
+	return nil
+}
+
+func (a *autoScaler) receiveLoop() {
+	producerMetrics := a.metricsReceiver.ProducerMetrics()
+	consumerMetrics := a.metricsReceiver.ConsumerMetrics()
+	for {
+		select {
+		case pm, ok := <-producerMetrics:
+			if ok {
+				a.receiveProducerMetric(pm)
+			}
+
+		case cm, ok := <-consumerMetrics:
+			if ok {
+				a.receiveConsumerMetric(cm)
+			}
+
+		case <-a.stop:
+			close(a.accumulatingStopped)
+			return
+		}
+	}
+}
+
+func (a *autoScaler) receiveConsumerMetric(cm metrics.ConsumerAggregateMetric) {
+	a.mutex.Lock()
+	defer a.mutex.Unlock()
+
+	funcTotals, ok := a.totals[cm.Topic]
+	if ok {
+		mt, ok := funcTotals[FunctionId{cm.ConsumerGroup}]
+		if ok {
+			mt.receiveCount += cm.Count
+		}
+	}
+}
+
+func (a *autoScaler) receiveProducerMetric(pm metrics.ProducerAggregateMetric) {
+	a.mutex.Lock()
+	defer a.mutex.Unlock()
+
+	funcTotals, ok := a.totals[pm.Topic]
+	if ok {
+		for _, mt := range funcTotals {
+			mt.transmitCount += pm.Count
+		}
+	}
+}
+
+type scaler func(*metricsTotals) int
+type adjuster func(int) int
+
+func decorate(s scaler, adjusters ...adjuster) scaler {
+	for _, a := range adjusters {
+		s = compose(a, s)
+	}
+
+	return s
+}
+
+func compose(a adjuster, s scaler) scaler {
+	return func(mt *metricsTotals) int {
+		return a(s(mt))
+	}
+}
+
+func (a *autoScaler) metricsScaler(fn FunctionId) func(mt *metricsTotals) int {
+	return func(mt *metricsTotals) int {
+		var proposedReplicas int
+		if mt.receiveCount == 0 {
+			if mt.transmitCount == 0 {
+				proposedReplicas = 0
+			} else {
+				proposedReplicas = 1 // arbitrary value
+			}
+		} else {
+			proposedReplicas = int(math.Ceil(float64(a.replicas[fn]) * float64(mt.transmitCount) / float64(mt.receiveCount)))
```

Comment:
I realize now that there is something I am not at ease with in this formula:
the proposed number of replicas should be the exact number that can cope with load (transmitCount), given prior knowledge of what functions are capable of handling (receiveCount). Why does it have to depend on the *current* number of replicas?

Reply:
Maybe it helps to think in terms of rates of production and consumption rather than counts. Now take an example. Suppose a functions is currently capable of consuming 1000 messages per second but the production rate has risen to 2000 per second, then we clearly have to double to rate of consumption. How can we determine the number of replicas needed without knowing the current number? Specifically, if there is currently one replica, then we should scale up to two. But if there are 100 replicas, we need to scale up to 200.

(A separate point which is not yet addressed is that we should propose just slightly more replicas than are needed because "a non-empty queue is an unstable queue" to quote a friend who works on a commercial message broker. In other words, if we don't drain a queue, we risk it becoming indefinitely large over time.)

Reply:
In your sentence above
> Take an example. Suppose a function is currently capable of consuming 1000 messages 

you shouldn't write *currently*. The consumption rate is the absolute consumption rate. Because if it isn't, then maybe we can cope with 2000/s but we don't know it (yet)

Other than that, I agree with your reasoning. But this worries me. What I hoped we could ultimately derive is how maning msg/s a function pod could cope with (assuming characteristics of such pods don't change, of course).

Reply:
I wrote "currently" because the absolute consumption rate (which is what we hope to measure once we have aggregated metrics) isn't a constant - it depends on the number of replicas. It may be best to chat about this if you are still worried. :-)

Reply:
I think I may be starting to understand your concern. The historical absolute consumption rate should eventually be gathered over a longer period of time and the number of replicas will vary over that period. We'll build up knowledge of consumption rate per pod and then can propose an absolute number of pods based on that rate and the current production rate, regardless of the current number of pods. On that basis, I think we have to wait and see until we get into the metrics aggregation side of things...